wolfcrypt/src/pkcs7.c

```diff
@@ -2345,10 +2426,282 @@ static int wc_PKCS7_SignedDataBuildSignature(PKCS7* pkcs7,
+                /* encrypt and flush out data */
+                switch (cipherType) {
+                    case WC_CIPHER_NONE:
+                        XMEMCPY(encContentOut, contentData, BER_OCTET_LENGTH);
+                        if (esd && esd->contentDigestSet != 1) {
+                            ret = wc_HashUpdate(&esd->hash, esd->hashType,
+                                contentData, BER_OCTET_LENGTH);
+                        }
+                        break;
+
+                #ifndef NO_AES
+                    case WC_CIPHER_AES_CBC:
+                        ret = wc_AesCbcEncrypt(aes, encContentOut,
+                            contentData, BER_OCTET_LENGTH);
+                        break;
+                #endif
+
+                #ifdef WOLFSSL_AESGCM_STREAM
+                    case WC_CIPHER_AES_GCM:
+                        ret = wc_AesGcmEncryptUpdate(aes, encContentOut,
+                                contentData, BER_OCTET_LENGTH, NULL, 0);
+                        break;
+                #endif
+                }
+
+            #ifdef WOLFSSL_ASYNC_CRYPT
+                /* async encrypt not available here, so block till done */
+                if (cipherType != WC_CIPHER_NONE) {
+                    ret = wc_AsyncWait(ret, &aes->asyncDev, WC_ASYNC_FLAG_NONE);
+                }
+            #endif
+                if (pkcs7->encodeStream) {
+                    wc_PKCS7_WriteOut(pkcs7,
+                                     (out)? out + outIdx: NULL,
+                                     encContentOutOct,
+                                     encContentOutOctSz);
+                    outIdx += encContentOutOctSz;
+                }
+                wc_PKCS7_WriteOut(pkcs7,
+                                    (out)? out + outIdx : NULL,
+                                    encContentOut,
+                                    BER_OCTET_LENGTH);
+                outIdx += BER_OCTET_LENGTH;
+
+                /* copy over any remaining data */
+                XMEMCPY(contentData, buf + sz, contentDataRead);
+                idx = contentDataRead;
+            }
+            else {
+                /* was not on an octet boundary, copy full
+                 * amount over */
+                XMEMCPY(contentData + idx, buf, sz);
+                idx += sz;
+            }
+        } while (totalSz < pkcs7->contentSz);
+
+        /* add in padding to the end */
+        if ((cipherType != WC_CIPHER_NONE) && (totalSz == pkcs7->contentSz)) {
+            int i;
+
+            if (BER_OCTET_LENGTH < idx) {
+                XFREE(encContentOut, heap, DYNAMIC_TYPE_PKCS7);
+                XFREE(contentData, heap, DYNAMIC_TYPE_PKCS7);
+                return BAD_FUNC_ARG;
+            }
+
+            for (i = 0; i < padSz; i++) {
+                contentData[idx + i] = (byte)padSz;
+            }
+            idx += padSz;
+        }
+
+
+        /* encrypt and flush out remainder of content data */
+        switch (cipherType) {
```

Comment:
This looks the same as in the loop except for the length being idx instead of BER_OCTET_LENGTH.
Extract out as a new function.